dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/EnrollmentSecurityValidationHook.java

```diff
@@ -0,0 +1,186 @@
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import org.hisp.dhis.organisationunit.OrganisationUnit;
+import org.hisp.dhis.program.Program;
+import org.hisp.dhis.program.ProgramInstance;
+import org.hisp.dhis.trackedentity.TrackedEntityInstance;
+import org.hisp.dhis.tracker.bundle.TrackerBundle;
+import org.hisp.dhis.tracker.domain.Enrollment;
+import org.hisp.dhis.tracker.preheat.PreheatHelper;
+import org.hisp.dhis.tracker.report.TrackerErrorCode;
+import org.hisp.dhis.tracker.report.TrackerErrorReport;
+import org.hisp.dhis.tracker.report.ValidationErrorReporter;
+import org.hisp.dhis.user.User;
+import org.springframework.stereotype.Component;
+
+import java.util.List;
+import java.util.Objects;
+
+import static org.hisp.dhis.tracker.report.ValidationErrorReporter.newReport;
+
+/**
+ * @author Morten Svanæs <msvanaes@dhis2.org>
+ */
+@Component
+public class EnrollmentSecurityValidationHook
+    extends AbstractTrackerValidationHook
+{
+
+    @Override
+    public int getOrder()
+    {
+        return 101;
+    }
+
+    @Override
+    public List<TrackerErrorReport> validate( TrackerBundle bundle )
+    {
+        ValidationErrorReporter reporter = new ValidationErrorReporter( bundle, this.getClass() );
+
+        User actingUser = bundle.getPreheat().getUser();
+
+        // NOTE: This should be handled at a higher level. But this is pretty paramount to security checks,
+        // conversion by convention like null==admin should be converted one place not on method input everywhere.
+        if ( actingUser == null )
+        {
+            throw new IllegalArgumentException( "User can't be null" );
+        }
+
+        for ( Enrollment enrollment : bundle.getEnrollments() )
+        {
+            reporter.increment( enrollment );
+
+            Program program = PreheatHelper.getProgram( bundle, enrollment.getProgram() );
+            OrganisationUnit organisationUnit = PreheatHelper.getOrganisationUnit( bundle, enrollment.getOrgUnit() );
+            TrackedEntityInstance trackedEntityInstance = PreheatHelper
+                .getTrackedEntityInstance( bundle, enrollment.getTrackedEntityInstance() );
+
+            // NOTE: maybe this should qualify as a hard break, on the prev hook (required properties).
+            if ( program == null || organisationUnit == null || trackedEntityInstance == null )
+            {
+                continue;
+            }
+
+            // See note below on this.
+            if ( !organisationUnitService.isInUserHierarchyCached( actingUser, organisationUnit ) )
+            {
+                reporter.addError( newReport( TrackerErrorCode.E1028 )
+                    .addArg( trackedEntityInstance )
+                    .addArg( program ) );
+            }
+
+            // This method "trackerOwnershipManager.hasAccess()" does a lot of things,
```

Comment:
@Bekkalizer chime in here? I am not sure how we want to deal with data write if you have, say, temporary access. Is it temporary read access or read and write?